research/delf/delf/python/training/global_features/train_utils.py

```diff
@@ -0,0 +1,368 @@
+# Lint as: python3
+# Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Training utilities for Global Features model."""
+
+import os
+import pickle
+import time
+
+import numpy as np
+import tensorflow as tf
+
+from delf.python import whiten
+from delf.python.datasets.revisited_op import dataset as test_dataset
+from delf.python.datasets.sfm120k import sfm120k
+from delf.python.training import global_features_utils
+from delf.python.training.model import global_model
+
+
+def grad(criterion, model, input, target, neg_num=5):
+  """Records gradients and loss through the network.
+
+  Args:
+    criterion: Loss function.
+    model: Network for the gradient computation.
+    input: Tuple of query, positive and negative images.
+    target: List of indexes to specify queries (-1), positives(1), negatives(0).
+    neg_num: Integer, number of negatives per a tuple.
+
+  Returns:
+    loss: Loss for the training step.
+    gradients: Computed gradients for the network trainable variables.
+  """
+  # Record gradients and loss through the network.
+  with tf.GradientTape() as tape:
+    output = tf.Variable(
+            tf.zeros(shape=(0, model.meta['outputdim']), dtype=tf.float32))
+    for img in input:
+      # Compute descriptor vector for each image.
+      o = model(tf.expand_dims(img, axis=0), training=True)
+      output = tf.concat([output, o], 0)
+
+    queries = tf.boolean_mask(output, target == -1, axis=0)
+    positives = tf.boolean_mask(output, target == 1, axis=0)
+    negatives = tf.boolean_mask(output, target == 0, axis=0)
+    negatives = tf.reshape(negatives, [tf.shape(queries)[0], neg_num,
+                                       model.meta['outputdim']])
+    # Loss calculation.
+    loss = criterion(queries, positives, negatives)
+
+  return loss, tape.gradient(loss, model.trainable_variables)
+
+
+def train_val(loader, model, criterion, optimizer, epoch, train=True,
```

Comment:
reading this fn, I believe it is doing training (or validation) for one entire epoch, right? If so, maybe the name could be changed to `train_val_one_epoch`